### DISCUSSION TITLE
Attitude & Rate Controller Gain question

### DISCUSSION THREAD

**Discussion by GNC-JY** [bitcraze/discussions#961] [Q&A] [answered]
Hello.

I'm curious about the Controller Gain value that you set on Crazyflie.

There should be a design criteria to set the Gain value, but I wonder how you set the Gain value.

In addition, it is judged that the output value from the controller loop through the rate control is distributed directly to the motor through the motor allocation, and the dc motor rotates. Normally, values such as thrust coefficient are multiplied and distributed to the motor output value, but I wonder if these values are included in setting the Gain value.

And explain why allocation is in the following matrix.
![image](https://github.com/bitcraze/discussions/assets/110527007/28d1a868-68d5-401e-950a-d2bd03cb291e)

best regards.

**Comment by knmcguire** [comment]
Hi! It would be helpful to point us to parts of the code that you need an explanation of. 

Also, there are 4 controllers on the crazyflie, are you referring to the PID controller?

Mind that the PID controller has been there for a while and that tuning has been done not by theory but by practice, so it could be that it hasn't been done the 'theoratical' way. The thrust and roll/pitch/yaw outputs of the controller goes into motor mixing equations first before being distributed to the motors.

**Reply by GNC-JY** [reply]
https://github.com/bitcraze/crazyflie-firmware/blob/cd3f3dfb668261f4461f972be8116d996aa7256e/src/modules/src/power_distribution_quadrotor.c#L79

Is it right to be distributed to the motor through this part? I checked this part.

Also, I am right to say PID controller.

Is the pid gain of the attitude / rate controller the value determined by experimenting with try & error?

Can you tell me where the motor mixing equation code is located?

**Reply by knmcguire** [reply, marked as answer]
Trial & error, yes. First the attitude controller existed, and those gains haven't changed in years. The position controller on top has also been tuned arbitrarily.

And actually, the code that you have given in the last post is the motor mixing. It transforms the roll pitch yaw and thrust commands to what needs to be distributed to each of the 4 motors. 

```
  motorThrustUncapped->motors.m1 = control->thrust - r + p + control->yaw;
  motorThrustUncapped->motors.m2 = control->thrust - r - p - control->yaw;
  motorThrustUncapped->motors.m3 = control->thrust + r - p + control->yaw;
  motorThrustUncapped->motors.m4 = control->thrust + r + p - control->yaw;

```
